### DISCUSSION TITLE
Some areas are not visible after I converted to standard A4 size, how to solve it? I want to get a full A4 pdf with correct page orientation and fully fill the A4 page without white margins.

### DISCUSSION THREAD

**Discussion by nissansz** [pymupdf/PyMuPDF#4083] [Q&A] [answered]
Original pdf is below
[p21.pdf](https://github.com/user-attachments/files/17892145/p21.pdf)

Converted is below
[p21_A4.pdf](https://github.com/user-attachments/files/17892158/p21_A4.pdf)
![image](https://github.com/user-attachments/assets/47c06e8e-3c8c-4ee7-92e3-9e26768165b1)
There is also white margin, how to remove it?

Code used.
'''
import fitz  # PyMuPDF

def convert_pdf_to_a4(input_pdf_path, output_pdf_path):
    # 定义 A4 尺寸（以点为单位，1英寸 = 72点）
    a4_width = 595.28  # A4 width in points
    a4_height = 841.89  # A4 height in points

    # 打开输入 PDF 文件
    pdf_document = fitz.open(input_pdf_path)

    # 创建一个新的 PDF 文件
    new_pdf_document = fitz.open()

    for page in pdf_document:
        # 获取原始页面的尺寸和旋转角度
        # original_rect = page.rect
        original_rect = page.mediabox
        # original_rect = page.get_pixmap(dpi=300)
        rotation = page.rotation
        print(rotation)

        # 根据旋转角度调整矩形的宽高
        if rotation == 0 or rotation == 180:  # 处理0度和180度
            rect_width, rect_height = original_rect.width, original_rect.height
        elif rotation == 90:  # 处理90度
            rect_width, rect_height = original_rect.height, original_rect.width
        elif rotation == 270:  # 处理270度
            rect_width, rect_height = original_rect.height, original_rect.width
        else:
            continue  # 如果有其他旋转值可以选择跳过

        # 创建一个新的 A4 页面
        a4_page = new_pdf_document.new_page(width=a4_width, height=a4_height)

        # 计算缩放比例，以使内容适应 A4 页面
        scale_x = a4_width / rect_width
        scale_y = a4_height / rect_height
        scale = min(scale_x, scale_y)  # 保持纵横比

        # 计算将页面图像居中放置的偏移量
        x_offset = (a4_width - rect_width * scale) / 2
        y_offset = (a4_height - rect_height * scale) / 2

        # 通过创建一个矩形区域并绘制页面
        scaled_rect = fitz.Rect(x_offset, y_offset, x_offset + rect_width * scale, y_offset + rect_height * scale)

        # 根据旋转角度应用旋转
        if rotation == 90:
            effective_rotation = -90
        elif rotation == 180:
            effective_rotation = -180
        elif rotation == 270:
            effective_rotation = -270
        else:
            effective_rotation = 0

        # 显示页面
        a4_page.show_pdf_page(scaled_rect, pdf_document, page.number, rotate=effective_rotation)

    # 保存新的 PDF
    new_pdf_document.save(output_pdf_path)
    new_pdf_document.close()
    pdf_document.close()

# 使用示例
input_pdf = r"F:\Binder1 - 副本.pdf"  # 输入文件路径
input_pdf = r"F:\p21.pdf"  # 输入文件路径
output_pdf = input_pdf.replace('.pdf', '') +'_A4.pdf' # 输出文件路径
convert_pdf_to_a4(input_pdf, output_pdf)
'''

**Comment by JorjMcKie** [comment, marked as answer]
The following 10-line script does the job:
```python
import sys, pymupdf

src = pymupdf.open(sys.argv[1])  # file to be reformatted to A4
out = pymupdf.open()  # result file
for spage in src:  # walk through input pages
    spage.remove_rotation()  # remove rotation - keeping visual appearance
    page = out.new_page()  # make A4 output page
    # insert source page content, appropriately scaled to full A4
    page.show_pdf_page(page.rect, src, spage.number)
out.ez_save(src.name + "-reformatted.pdf")
```

Because your example file pages have an aspect ratio different from sqrt(2):1, white borders result left and right. Specify `keep_proportion=False` to stretch.

**Reply by nissansz** [reply]
Great, it works. Thanks.

**Comment by nissansz** [comment]
Another question is: original pdf is very small,  241KB.
But if I convert it with pixelmap (like 300 dpi, 600 dpi), ->cv2, or pillow, then resave to a new pdf, the size may become very large.
How to match original size with get pixelmap?

**Reply by JorjMcKie** [reply]
My above script keeps or even reduces the original PDF size:
![image](https://github.com/user-attachments/assets/5f2206ca-d01b-41f3-a5b4-0a760040042b)

The embedded image at xref 12 is a JPEG image (RGB). If you insert a pixmap the resulting image format is PNG - which behaves differently under compression.
This is presumably also true with cv2 images.
If you change resolution, then that will also cause different compression behaviors
If no transparency is involved, you can use the JPEG output of a pixmap (`pix.tobytes("jpeg")`) and insert this via the `stream` parameter. Respectively with cv2 images I suppose.

In any case, do you maximum PDF compression on output (ez_save). There exists a new parameter `compression_effort` which may be set to 100 (maximum compression): sometimes this also has effect.

**Comment by nissansz** [comment]
Thank you.
I just want to find a way to minimize image size (bust same quality as original) when exported to jpg or png from pdf,  and make some edits to images,
and then recreate pdf from edited images, to meet same pdf size and same image quality as original pdf.

Any possible solution?

**Reply by nissansz** [reply]
Save to pptx, then editing in pptx, then print to pdf,
this way seems working.
Maybe I should insert line, text etc, directly in pdf without exporting images.

**Reply by JorjMcKie** [reply]
I would always use the in-built export to PDF features of the office software instead of print-to-pdf.
Because MS Office / LibreOffice know what they are dealing with in these cases and thus are more specialized.
Especially LibreOffice: my favorite. Always better, smaller an faster than the MS alternative.

**Comment by nissansz** [comment]
There are 2 column pages,
I want to convert 1 page to 2 pages, and 1 page is left only, the other page is leftover parts.
What is the best option?

Use Insert_rect for white textbox to cover?

**Reply by JorjMcKie** [reply]
You can use the "clip" parameter in `show_pdf_page()`. This will suppress the display of stuff outside.
But only "suppress display": things will remain extractable in entirety.
You can use redactions to actually remove unwanted things.

**Reply by nissansz** [reply]
show_pdf_page(rect, docsrc, pno=0, keep_proportion=True, overlay=True, oc=0, rotate=0, clip=None)

It works

'''
        for box in bounding_boxes[1:]: #去掉左侧的元素
            left = box[0]
            top = box[1]
            right = box[0] + box[2]
            bottom = box[1] + box[3]

            clip = fitz.Rect(left, top, right, bottom)
            page.show_pdf_page(clip , src, pno=i, keep_proportion=True, overlay=True, oc=0, rotate=0,
                               clip=clip)
'''

**Reply by nissansz** [reply]
Is there any option to reverse show clip?
For example. the left boxes are clip on page 1, on page 2, I want to hide all parts on page 1

**Reply by JorjMcKie** [reply]
No. Just use a different clip.

**Reply by nissansz** [reply]
When I use below code, the newly drawn rect cannot cover the areas, any method to cover?
I can only see a yellow area in a second, then it disappeared.

'''
    for box in leftboxes:  # 去掉左侧的元素
        clip = fitz.Rect(box[0], box[1], box[2], box[3])
        page.show_pdf_page(page.rect, src, pno=i, keep_proportion=True, overlay=True, oc=0, rotate=0,
                           clip=None)
        # page.draw_rect(clip, fill=(1, 1, 1), color=None)
        page.draw_rect(clip, color=None, fill=fitz.pdfcolor["yellow"], fill_opacity=0.5, overlay=True)
'''

**Reply by JorjMcKie** [reply]
Shouldn't happen.

**Reply by nissansz** [reply]
[2024-11-27_18-20-52.zip](https://github.com/user-attachments/files/17932775/2024-11-27_18-20-52.zip)
Check above video, in the middle, there is yellow.

This is the generated pdf.
[JP2000113653A.PDF_split.pdf](https://github.com/user-attachments/files/17932782/JP2000113653A.PDF_split.pdf)

**Reply by nissansz** [reply]
place  page.show_pdf_page( before for loop, it can work.
'''
 page.show_pdf_page(page.rect, src, pno=i, keep_proportion=True, overlay=True, oc=0, rotate=0,
                       clip=None)
    # for box in bounding_boxes:
    for box in leftboxes:  # 去掉左侧的元素
        clip = fitz.Rect(box[0], box[1], box[2], box[3])

        # # page.draw_rect(clip, fill=(1, 1, 1), color=None)
        page.draw_rect(clip, color=None, fill=fitz.pdfcolor["yellow"], fill_opacity=0.5, overlay=True)

'''

**Comment by nissansz** [comment]
Can redactions remove rect any area to be blank?

**Reply by nissansz** [reply]
redaction = page.add_redact_annot(rect, fill=(1, 1, 1), text="")
        page.apply_redactions()

I can remove area by above code,
comparing with draw_rect, redct result pdf size is larger.

Another problem is that, when I wrote my own ocr text html layer to original pdf, (used this option doc.ez_save(new_file, garbage=3, deflate=True))
the pdf file is very slow to browse, the more pages, the slower, 
Is there any parameters to accelerate during save, draw, etc.?

See below example.

[JP2000113653A_300_ocr.pdf](https://github.com/user-attachments/files/17933319/JP2000113653A_300_ocr.pdf)
